### ISSUE TITLE
Ability to store objects created externally

### ISSUE THREAD

**Issue by drekka** [Swinject/Swinject#204]
For example, I might create a user object as a result of a server call, then want Swinject to store it in a singleton so I can inject it elsewhere.

So something like:

    let user:User = //...
    let registration = container.findRegistration(User.self)
    registration.store(user)

**Comment by jakubvano** [comment]
Hi @drekka, 
What you are describing cannot (should ?) not be solved by Swinject alone, but I would suggest something like:
```swift
protocol UserStorage {
 var user: User? { get set }
}

class NetworkingService {
   init(userStorage: UserStorage) { ... }
   ...
  private func handleResponse() {
     ...
     userStorage.user = obtainedUser
  }
}
```
And wherever you have dependency on user, you would change it to dependency on `UserStorage`.

Hope this helps 😉

**Comment by drekka** [comment]
If understand your response, you're suggesting I should create a wrapper for such data that is instantiated by Swinject. That would work. But it seems like a less than great solution. This scenario is not something I do all the time. But I would have to create wrappers for every type of object I wanted Swinject to manage for me.

**Comment by jakubvano** [comment]
I get what you are trying to do, but IMO it is something that should not be done via DI framework - DI frameworks are supposed to be used for managing "Data manipulators" (i.e. objects that deal with data) not "Data" directly.

Of course, you can try bending Swinject to cover this usage, something like
```swift
class NetworkingService {
   init(container: Container) { ... }

   private func handleResponse() {
      ...
     // This would "override" previous registrations for `User`, so all subsequent
     // container.resolve(User.self) would return `obtainedUser`
      container.register(User.self) { _ in obtaineduser } 
   }
}
```
but you migth run into some unexpected behavior

**Comment by yoichitgy** [comment]
Learning [Service Locator pattern](https://msdn.microsoft.com/en-us/library/ff921142.aspx) might be helpful. A [DI](https://msdn.microsoft.com/en-us/library/ff921152.aspx) container can be used in the pattern, although some people mention it as an anti pattern.
http://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/

(I thought @drekka's question was about [runtime arguments](https://github.com/Swinject/Swinject/blob/master/Documentation/DIContainer.md#registration-with-arguments-in-a-di-container) when I saw the question at the first time.)

**Comment by drekka** [comment]
Creating a container which can be used to store a value will solve the problem. Effectively also what @yoichitgy is saying by talking about the SLP. Regardless of whether it's a simple wrapper or a locator, it still feels not great :-) having to wrap an object just so I can replace it. 

Another scenario is one I'm wrestling with at the moment - connections. I have an object that represents and manages a connection to a server. If I loose the connection to the server or want to use a different server I need to re-create this object. To keep things nicely decoupled and testable I want to have the DI container manage this object and inject it where needed. Unfortunately none of Swinject's scopes address the idea of being able to store an object which is regarded as a singleton, and then set a new instance at a later stage. So again a wrapper will solve this even though it's not as simple or as nice as simply being able to update the stored object.

**Comment by jakubvano** [comment]
About your second issue - `2.0.0-beta.3` provides ability to create [custom object scopes](https://github.com/Swinject/Swinject/blob/master/Documentation/ObjectScopes.md#custom-scopes), so maybe it will help you.
As for replacing a singleton - you can possibly destroy and recreate an instance managed by `Container`, but issue is with other objects that have a reference to this instance - I see 2 options:

 - recreate those objects together with connection service: custom object scope will help with this, but you will probably need to adjust your app to support this (what about objects referencing objects referencing objects..? and eventually `UIViewController`s which are retained by `UIKit`..)

- create a wrapper which allows you to access connection service dynamically

**Comment by yoichitgy** [comment]
@drekka, was the issue resolved?

**Comment by drekka** [comment]
Sort of. I ended up writing classes that wrapped optional values and instantiating them with SwInject. For example:

    class CurrentUser {
        var user:User?
    }
    class ServerConnection {
        var server:Server?
    }
    // etc


This sort of structure allows the app to handle users logging in and out, whilst SWInject can inject the wrapping `CurrentUser` instance.

As @jakubvano says, SWinject is not for managing data. Which I agree with in the context of data. i.e. I would not use a DI framework to store records from a database. 

However there are sometimes high level instances of things (users, network connections, etc) which are part of the app rather than data, and are by nature transient, but still need to be injected.

Before switching to Swift and discovering SWInject, I wrote a DI framework for Objective-C (https://github.com/drekka/Alchemic). In that framework I included several concepts such as nillable injections where a nil value was a valid injection, and transient managed instances where the instance being managed by the framework could change over time with the framework atomically updating all references to it.  

So situations where top level objects change periodically was where I was coming from in my original request.

Writing a DI framework is a very interesting and challenging project to take on and I wouldn't expect SWInject to do the same sort of things I did with mine. Particularly as Swift and Objective-C provide very different options and runtimes.

But it would be nice if I didn't have to write wrappers such as the above. Perhaps when I get some time I'll look into a custom object scope and see what I can do with it.
